[Flake] Unit Test: TestControllerGameServerCount

<!-- Please use this template while reporting a bug and provide as much info as possible. Not doing so may result in your bug not being addressed in a timely manner. Thanks!

If the matter is security related, please disclose it privately via 
-->


**What happened**:

Flake:

```
{"message":"Wait for cache sync","severity":"info","time":"2022-11-10T23:29:25.722243823Z"}
{"message":"unable to calculate 'Creating' state duration of '89hzzgz67v' GameServer","severity":"warning","source":"*metrics.Controller","time":"2022-11-10T23:29:25.828551589Z"}
{"message":"unable to calculate 'Ready' state duration of '89hzzgz67v' GameServer","severity":"warning","source":"*metrics.Controller","time":"2022-11-10T23:29:26.829692585Z"}
--- FAIL: TestControllerGameServerCount (3.11s)
    controller_test.go:76: 
        	Error Trace:	controller_test.go:76
        	            				controller_test.go:146
        	Error:      	Not equal: 
        	            	expected: 3
        	            	actual  : 4
        	Test:       	TestControllerGameServerCount
        	Messages:   	number of timeseries does not match under metric: gameservers_count
    controller_test.go:83: 
        	Error Trace:	controller_test.go:83
        	            				controller_test.go:146
        	Error:      	Should be true
        	Test:       	TestControllerGameServerCount
        	Messages:   	no TimeSeries found with labels: [test-fleet default Creating]
    controller_test.go:84: 
        	Error Trace:	controller_test.go:84
        	            				controller_test.go:146
        	Error:      	Not equal: 
        	            	expected: []string(nil)
        	            	actual  : []string{"test-fleet", "default", "Creating"}
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1,2 +1,6 @@
        	            	-([]string) <nil>
        	            	+([]string) (len=3) {
        	            	+ (string) (len=10) "test-fleet",
        	            	+ (string) (len=7) "default",
        	            	+ (string) (len=8) "Creating"
        	            	+}
        	            	 
        	Test:       	TestControllerGameServerCount
        	Messages:   	label values don't match
    controller_test.go:86: 
        	Error Trace:	controller_test.go:86
        	            				controller_test.go:146
        	Error:      	Not equal: 
        	            	expected: <nil>(<nil>)
        	            	actual  : int64(0)
        	Test:       	TestControllerGameServerCount
        	Messages:   	metric: gameservers_count, tags: [{test-fleet true} {default true} {Creating true}], values don't match; got: 0, want: <nil>
{"message":"Wait for cache sync","severity":"info","time":"2022-11-10T23:29:28.833974642Z"}
{"message":"Wait for cache sync","severity":"info","time":"2022-11-10T23:29:30.939633368Z"}
```

**What you expected to happen**:

The test to pass always

**How to reproduce it (as minimally and precisely as possible)**:

Run the test lots.

**Anything else we need to know?**:

It mostly passes!

**Environment**:
- Agones version: dev
- Kubernetes version (use `kubectl version`): N/A
- Cloud provider or hardware configuration: N/A
- Install method (yaml/helm): N/A
- Troubleshooting guide log(s): See above.
- Others: N/A

--- a/pkg/metrics/controller_test.go
+++ b/pkg/metrics/controller_test.go
@@ -25,6 +25,7 @@ import (
 	agtesting "agones.dev/agones/pkg/testing"
 	"agones.dev/agones/pkg/util/runtime"
 	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
 	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
 	"go.opencensus.io/metric/metricdata"
@@ -100,14 +101,21 @@ func TestControllerGameServerCount(t *testing.T) {
 	c := newFakeController()
 	defer c.close()
 
+	c.run(t)
+	require.True(t, c.sync())
+
 	gs1 := gameServerWithFleetAndState("test-fleet", agonesv1.GameServerStateCreating)
 	c.gsWatch.Add(gs1)
+	require.Eventually(t, func() bool {
+		gs, err := c.gameServerLister.GameServers(gs1.ObjectMeta.Namespace).Get(gs1.ObjectMeta.Name)
+		assert.NoError(t, err)
+		return gs.Status.State == agonesv1.GameServerStateCreating
+	}, 5*time.Second, time.Second)
+	c.collect()
+
 	gs1 = gs1.DeepCopy()
 	gs1.Status.State = agonesv1.GameServerStateReady
 	c.gsWatch.Modify(gs1)
-
-	c.run(t)
-	require.True(t, c.sync())
 	require.Eventually(t, func() bool {
 		gs, err := c.gameServerLister.GameServers(gs1.ObjectMeta.Namespace).Get(gs1.ObjectMeta.Name)
 		assert.NoError(t, err)
@@ -121,7 +129,6 @@ func TestControllerGameServerCount(t *testing.T) {
 	c.gsWatch.Add(gameServerWithFleetAndState("", agonesv1.GameServerStatePortAllocation))
 	c.gsWatch.Add(gameServerWithFleetAndState("", agonesv1.GameServerStatePortAllocation))
 
-	c.run(t)
 	require.True(t, c.sync())
 	// Port allocation is last, so wait for that come to the state we expect
 	require.Eventually(t, func() bool {
@@ -131,10 +138,11 @@ func TestControllerGameServerCount(t *testing.T) {
 
 		for _, m := range ex.metrics {
 			if m.Descriptor.Name == gameServersCountName {
-				for _, d := range m.TimeSeries {
-					if d.LabelValues[0].Value == "none" && d.LabelValues[1].Value == defaultNs && d.LabelValues[2].Value == "PortAllocation" {
-						return d.Points[0].Value == int64(2)
-					}
+				if len(m.TimeSeries) == 4 {
+					return true
+				} else {
+					logrus.WithField("m", m).Info("Metrics")
+					return false
 				}
 			}
 		}
@@ -144,6 +152,7 @@ func TestControllerGameServerCount(t *testing.T) {
 
 	reader.ReadAndExport(exporter)
 	assertMetricData(t, exporter, gameServersCountName, []expectedMetricData{
+		{labels: []string{"test-fleet", defaultNs, "Creating"}, val: int64(0)},
 		{labels: []string{"test-fleet", defaultNs, "Ready"}, val: int64(0)},
 		{labels: []string{"test-fleet", defaultNs, "Shutdown"}, val: int64(1)},
 		{labels: []string{"none", defaultNs, "PortAllocation"}, val: int64(2)},

--- a/pkg/metrics/util_test.go
+++ b/pkg/metrics/util_test.go
@@ -17,6 +17,7 @@ package metrics
 import (
 	"context"
 	"reflect"
+	"strconv"
 	"testing"
 
 	agonesv1 "agones.dev/agones/pkg/apis/agones/v1"
@@ -32,6 +33,11 @@ import (
 	"k8s.io/apimachinery/pkg/watch"
 	k8stesting "k8s.io/client-go/testing"
 	"k8s.io/client-go/tools/cache"
+)
+
+var (
+	// counter for unique GameServer names
+	counter = 0
 )
 
 // newFakeController returns a controller, backed by the fake Clientset
@@ -124,9 +130,11 @@ func gameServerWithFleetAndState(fleetName string, state agonesv1.GameServerStat
 	if fleetName != "" {
 		lbs[agonesv1.FleetNameLabel] = fleetName
 	}
+	// ensure the name is unique each time.
+	counter++
 	gs := &agonesv1.GameServer{
 		ObjectMeta: metav1.ObjectMeta{
-			Name:      rand.String(10),
+			Name:      rand.String(10) + strconv.Itoa(counter),
 			Namespace: "default",
 			UID:       uuid.NewUUID(),
 			Labels:    lbs,

